### PR TITLE
Changes for enforce_settings to enforce run_settings over cypress conf

### PR DESCRIPTION
**Pull request by hariharanbrowserstack**
Description:
- CLI changes for enforcing run_settings in browserstack.json over cypress config file.

Jira:
- https://browserstack.atlassian.net/browse/APS-8909

### CODE REVIEW

File: bin/helpers/capabilityHelper.js [hunk 1]
```diff
@@ -234,6 +239,9 @@ const validate = (bsConfig, args) => {
         // Detect if the user is not using the right directory structure, and throw an error
         if (!Utils.isUndefined(cypressConfigFile.integrationFolder) && !Utils.isCypressProjDirValid(bsConfig.run_settings.cypressProjectDir,cypressConfigFile.integrationFolder)) reject(Constants.validationMessages.INCORRECT_DIRECTORY_STRUCTURE);
       }
+      logger.debug("Validating baseurl and integrationFolder in browserstack.json");
+      if (!Utils.isUndefinedOrFalse(bsConfig.run_settings.enforce_settings) &&  !Utils.isUndefined(bsConfig.run_settings.baseUrl) && bsConfig.run_settings.baseUrl.includes("localhost") && !Utils.getLocalFlag(bsConfig.connection_settings)) reject(Constants.validationMessages.LOCAL_NOT_SET.replace("<baseUrlValue>", bsConfig.run_settings.baseUrl));
+      if (!Utils.isUndefinedOrFalse(bsConfig.run_settings.enforce_settings) &&  !Utils.isUndefined(bsConfig.run_settings.integrationFolder) && !Utils.isCypressProjDirValid(bsConfig.run_settings.cypressProjectDir,bsConfig.run_settings.integrationFolder)) reject(Constants.validationMessages.INCORRECT_DIRECTORY_STRUCTURE);
```
Comment (asambstack) [inline bin/helpers/capabilityHelper.js:244]: Can be added to else block for `Utils.isUndefinedOrFalse(bsConfig.run_settings.enforce_settings)`

File: bin/helpers/constants.js [hunk 2]
```diff
@@ -70,8 +70,12 @@ const userMessages = {
   UPLOADING_TESTS_SUCCESS: "Uploaded tests successfully",
   UPLOADING_NPM_PACKAGES: "Uploading required node_modules to BrowserStack",
   UPLOADING_NPM_PACKAGES_SUCCESS: "Uploaded node_modules successfully",
+  SKIP_CYPRESSCONFIG_INSTALL:
```
Comment (asambstack) [inline bin/helpers/constants.js:73]:
```suggestion
  SKIP_CYPRESS_CONFIG_INSTALL:
```

Comment (asambstack) [reply bin/helpers/constants.js:73]: Also, don't see where this constant is being used.

File: bin/helpers/utils.js [hunk 3]
```diff
@@ -1286,6 +1290,32 @@ exports.setConfig = (bsConfig, args) => {
   }
 }
 
+// set configs if enforce_settings is passed
+exports.setEnforceSettingsConfig = (bsConfig) => {
+  let config_args = (!this.isUndefined(bsConfig) && !this.isUndefined(bsConfig.run_settings) && !this.isUndefined(bsConfig.run_settings.config)) ? bsConfig.run_settings.config : undefined;
```
Comment (asambstack) [inline bin/helpers/utils.js:1295]:
```suggestion
  let config_args = (bsConfig && bsConfig.run_settings && bsConfig.run_settings.config) ? bsConfig.run_settings.config : undefined;
```

File: bin/helpers/utils.js [hunk 4]
```diff
@@ -1286,6 +1290,32 @@ exports.setConfig = (bsConfig, args) => {
...
+  if ( this.isUndefined(config_args) || !config_args.includes("video") ) {
```
Comment (asambstack) [inline bin/helpers/utils.js:1296]: Condition for `bsConfig.run_settings` might be required here I believe?

File: bin/helpers/utils.js [hunk 5]
```diff
@@ -1286,6 +1290,32 @@ exports.setConfig = (bsConfig, args) => {
...
+    let video_args = (this.isUndefined(bsConfig.run_settings.video_config) || this.isUndefined(bsConfig.run_settings.video_config.video) || !bsConfig.run_settings.video_config.video ) ? 'video=false' : 'video=true' ;
+    video_args += (this.isUndefined(bsConfig.run_settings.video_config) || this.isUndefined(bsConfig.run_settings.video_config.videoUploadOnPasses) || !bsConfig.run_settings.video_config.videoUploadOnPasses ) ? ',videoUploadOnPasses=false' : ',videoUploadOnPasses=true';
+    config_args = this.isUndefined(config_args) ? video_args : config_args + ',' + video_args;
+    logger.debug(`Setting video_args for enforce_settings to ${video_args}`);
+  }
+  if ( (!this.isUndefined(bsConfig) && !this.isUndefined(bsConfig.run_settings) && !this.isUndefined(bsConfig.run_settings.baseUrl)) && (this.isUndefined(config_args) || !config_args.includes("baseUrl")) ) {
```
Comment (asambstack) [inline bin/helpers/utils.js:1302]:
```suggestion
  if ( (bsConfig && bsConfig.run_settings && bsConfig.run_settings.baseUrl) && (this.isUndefined(config_args) || !config_args.includes("baseUrl")) ) {
```

File: bin/helpers/utils.js [hunk 6]
```diff
@@ -1286,6 +1290,32 @@ exports.setConfig = (bsConfig, args) => {
...
+  }
+  // set specs in config of specpattern to override cypress config
+  if(!this.isUndefined(bsConfig) && !this.isUndefined(bsConfig.run_settings) && !this.isUndefined(bsConfig.run_settings.specs)) {
+    let spec_pattern_args = 'specPattern="'+bsConfig.run_settings.specs+'"';
+    if( bsConfig.run_settings.cypressTestSuiteType !== Constants.CYPRESS_V10_AND_ABOVE_TYPE) {
+      spec_pattern_args = 'testFiles="'+bsConfig.run_settings.specs+'"';
+    }
+    config_args = this.isUndefined(config_args) ? spec_pattern_args : config_args + ',' + spec_pattern_args;
+  }
+  if ( !this.isUndefined(config_args) ) bsConfig["run_settings"]["config"] = config_args;
```
Comment (asambstack) [inline bin/helpers/utils.js:1315]: Probable can create a helper method `const isDefined = val => !this.isUndefined(val)` and use it everywhere

**Review by asambstack — CHANGES_REQUESTED** [review]

File: bin/helpers/utils.js (hunk 4 above)
Comment (hariharanbrowserstack) [reply bin/helpers/utils.js:1296]: This method is called only upon run_settings.enforce_settings is true. Hope that check wont be needed.

**Review by asambstack — DISMISSED** [review]

**Pull request merged**